Restructure GDB debugging and add guide for STLink and link to emdbg (#2651)
PX4/PX4-user_guide: en/debug/probe_stlink.md

# STLink Debug Probe

> The [STLinkv3-MINIE][stlinkv3] is a cheap, fast and highly capable debug probe that can serve as a stand-alone debug and console communicator for a PX4 developer:

Reviewer: ```suggestion
The [STLinkv3-MINIE](https://www.st.com/en/development-tools/stlink-v3minie.html) is a cheap, fast and highly capable debug probe that can serve as a stand-alone debug and console communicator for a PX4 developer:
```

:::note
There are no "out of the box" connectors for using the STLink Debug Probe with Pixhawk flight controllers.
The [Pixhawk Debug Port Adapter](#pixhawk-debug-port-adapter) section below explains how you can create your own (some soldering required).
:::

## Debugging Configuration

The STLink provides the [GDB server via OpenOCD](https://openocd.org/doc-release/html/index.html):

```sh
# Ubuntu
sudo apt install openocd
# macOS
brew install open-ocd
```

You can launch the GDB server in a new terminal shell:

```sh
openocd -f interface/stlink.cfg -f target/stm32f7x.cfg
```

The config file needs to be:

- FMUv2-v4: `-f target/stm32f4x.cfg`
- FMUv5: `-f target/stm32f7x.cfg`
- FMUv6: `-f target/stm32h7x.cfg`

You can then connect to port 3333 via GDB:

```sh
arm-none-eabi-gdb build/px4_fmu-v5x_default/px4_fmu-v5x_default.elf -ex "target extended-remote :3333"
```

See the [Embedded Debug Tools][emdbg] for more advanced debug options.

## Pixhawk Debug Port Adapter

To connect to the Pixhawk Debug Port of your PX4 Autopilot, you need to solder an adapter.

For this solder guide you need:

> - 1x [STLinkv3-MINIE][stlinkv3].

Reviewer: ```suggestion
- 1x [STLinkv3-MINIE](https://www.st.com/en/development-tools/stlink-v3minie.html).
```

> - 1x cable connector for mating with [JST SM10B (Full)][jst10] or [JST SM06B (Mini)][jst6].

Reviewer: ```suggestion
- 1x cable connector for mating with [JST SM10B (Full)](https://www.digikey.com/products/en?keywords=A10SR10SR30K203A) or [JST SM06B (Mini)](https://www.digikey.com/products/en?keywords=A06SR06SR30K152A).

```

  We recommend buying fully assembled cables with two connectors on either side.
- 1x soldering iron and solder.
- Some tongs, cutting pliers, and tweezers.

> The [Pixhawk Debug Port is standardized in DS-009][ds009] and needs to be connected to the STLinkv3-MINIE Board-To-Board (BTB) card edge connector CN2.

Reviewer: ```suggestion
The [Pixhawk Debug Port is standardized in DS-009](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-009%20Pixhawk%20Connector%20Standard.pdf) and needs to be connected to the STLinkv3-MINIE Board-To-Board (BTB) card edge connector CN2.
```

Alternatively, you can solder these connector pin off the PCB, but it can take longer.

![](../../assets/debug/stlinkv3_minie_p10.jpeg)

Rotate the STLinkv3 to cut off the other row, again being very careful to not damage it.

![](../../assets/debug/stlinkv3_minie_p11.jpeg)

Cut a ~5cm (~2in) long piece of shrink tube.
It should be flush with the USB-C connector and extend a little beyond the end.

![](../../assets/debug/stlinkv3_minie_p12.jpeg)

Hold both the PCB and the shrink tube with the flat tongs by the **bottom** metal part of the USB-C connector.
Be careful not to accidentally squeeze the middle plastic part of the USB-C connector!

![](../../assets/debug/stlinkv3_minie_p13.jpeg)

Use the heat gun to shrink the tubing all around the debug probe.
Make sure the tubing is equally shrunk and protects the whole PCB.

![](../../assets/debug/stlinkv3_minie_p14.jpeg)

Optionally, you may add a logo of your choice printed on paper and cut to size.
Be aware that the heat can make the ink flow a little, so you may need to experiment with what settings work with your printer.

![](../../assets/debug/stlinkv3_minie_p15.jpeg)

> [stlinkv3]: https://www.st.com/en/development-tools/stlink-v3minie.html

Reviewer: ```suggestion
```

> [ds009]: https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-009%20Pixhawk%20Connector%20Standard.pdf

Reviewer: ```suggestion
```

[emdbg]: https://pypi.org/project/emdbg/

> [jst6]: https://www.digikey.com/products/en?keywords=A06SR06SR30K152A

Reviewer: ```suggestion
```

> [jst10]: https://www.digikey.com/products/en?keywords=A10SR10SR30K203A

Reviewer: ```suggestion
```

![](../../assets/debug/stlinkv3_minie_p11.jpeg)

Cut a ~5cm (~2in) long piece of shrink tube.
It should be flush with the USB-C connector and extend a little beyond the end.

![](../../assets/debug/stlinkv3_minie_p12.jpeg)

Hold both the PCB and the shrink tube with the flat tongs by the **bottom** metal part of the USB-C connector.
Be careful not to accidentally squeeze the middle plastic part of the USB-C connector!

![](../../assets/debug/stlinkv3_minie_p13.jpeg)

Use the heat gun to shrink the tubing all around the debug probe.
Make sure the tubing is equally shrunk and protects the whole PCB.

![](../../assets/debug/stlinkv3_minie_p14.jpeg)

Optionally, you may add a logo of your choice printed on paper and cut to size.
Be aware that the heat can make the ink flow a little, so you may need to experiment with what settings work with your printer.

![](../../assets/debug/stlinkv3_minie_p15.jpeg)

[stlinkv3]: https://www.st.com/en/development-tools/stlink-v3minie.html
[ds009]: https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-009%20Pixhawk%20Connector%20Standard.pdf
[emdbg]: https://pypi.org/project/emdbg/

> [openocd]: https://openocd.org

Reviewer: Removing - only one link
```suggestion
```

> [stlink]: https://www.st.com/en/development-tools/stsw-link004.html

Reviewer: ```suggestion
```

# STLink Debug Probe

The [STLinkv3-MINIE][stlinkv3] is a cheap, fast and highly capable debug probe that can serve as a stand-alone debug and console communicator for a PX4 developer:

- Just one single USB-C connection for Reset, SWD, SWO, and serial in a very small package!
- Up to 24MHz SWD and SWO connection. Up to 16 MBaud serial. 1.65V to 3.6V target voltage. USB2 high-speed 480 Mbps connection.
- Driven by STLink or OpenOCD software with wide device support.
- Much cheaper (<15€) than a Pixhawk Debug Adapter (~20€) with a JLink EDU mini (~55€) or JLink BASE (~400€) while having better hardware specs.

:::note
There are no "out of the box" connectors for using the STLink Debug Probe with Pixhawk flight controllers.
The [Pixhawk Debug Port Adapter](#pixhawk-debug-port-adapter) section below explains how you can create your own (some soldering required).
:::

> ## Debugging Configuration

Reviewer: Soldering is a bit off-putting for many, and affects cost - its a lot more to use this if you don't have soldering kit/experience. I've added this as a note above, and separated the usage instructions under this heading "Debugging Configuration" - better name welcome!

We recommend soldering the long cable first and testing how fast you can communicate with your target.

:::note
This guide is written for the full 10-pin debug port.
If you want to solder the mini 6-pin version, just leave out the signals you don't have.
The STLink supports any SWD/JTAG-based debug interface, so you can adapt this guide for any other connector you may have.
The debug probes are so cheap, you can just have one per connector instead of using adapters.
:::

This is how the STLinkv3-MINIE is delivered.

![](../../assets/debug/stlinkv3_minie_p1.jpeg)

Unwrap the PCB and check it for any damage.
Plug it in and see if it powers on correctly.

![](../../assets/debug/stlinkv3_minie_p2.jpeg)

### Short Cable

The short cable requires a wire cutter and stripper and requires a little more soldering skill.
However, it makes the entire debug probe even smaller.

Assemble a 10-pin connector without GPIO1/2. If you already have an assembled cable, carefully remove the two GPIO1/2 cables with a tweezer by lifting the pegs that keep the cables secured.
Cut the cables to a short ~2cm (~1in) length and strip the wires.

> ![](../../assets/debug/stlinkv3_minie_p3.jpeg)

Reviewer: Instructions look good to me. FYI I losslessly shrunk the images.

Then solder the TX and RX pins. Solder the RST connection last.

![](../../assets/debug/stlinkv3_minie_p5.jpeg)

Turn the STLink over and solder the remaining three wires.
Start with SWDIO->TMS, then SWCLK->CLK, and finally SWO->TDO.

![](../../assets/debug/stlinkv3_minie_p6.jpeg)

### Long Cable

The long cable is particularly useful if you use pre-assembled cables as it removes the need to cut wires or strip them.

Carefully remove the two GPIO1/2 cables from one connector of the cable.
Then remove all cables from the other connector.
You are left with a eight crimped connectors at the end of the wires.

![](../../assets/debug/stlinkv3_minie_p7.jpeg)

Tin the crimped cable connectors and BTB connector and solder the crimped connectors directly to the STLinkv3.
Be careful to not create shorts between the cables, as the crimped connectors are quite large.

![](../../assets/debug/stlinkv3_minie_p8.jpeg)

### Testing

You should now test your debug probe to ensure you do not have any electrical shorts.

1. Plug the probe into your target via the Pixhawk Debug Port.
2. Test the serial port with a program of your choice.

> 3. Test the SWD and RST connection via [OpenOCD][] or [STLink][] software.

Reviewer: Missing references

Reviewer: Empty references - linked up using your references at the end, which only had one item.

```suggestion
3. Test the SWD and RST connection via [OpenOCD][https://openocd.org] or [STLink](https://www.st.com/en/development-tools/stsw-link004.html) software.
```

> 4. Test the SWO connection via [Orbuculum][].

Reviewer: Missing reference  - again, better to use the actual URL usually.

Reviewer: ```suggestion
4. Test the SWO connection via [Orbuculum][https://github.com/orbcode/orbuculum].
```

Alternatively, you can solder these connector pin off the PCB, but it can take longer.

![](../../assets/debug/stlinkv3_minie_p10.jpeg)

Rotate the STLinkv3 to cut off the other row, again being very careful to not damage it.

![](../../assets/debug/stlinkv3_minie_p11.jpeg)

Cut a ~5cm (~2in) long piece of shrink tube.
It should be flush with the USB-C connector and extend a little beyond the end.

![](../../assets/debug/stlinkv3_minie_p12.jpeg)

Hold both the PCB and the shrink tube with the flat tongs by the **bottom** metal part of the USB-C connector.
Be careful not to accidentally squeeze the middle plastic part of the USB-C connector!

![](../../assets/debug/stlinkv3_minie_p13.jpeg)

Use the heat gun to shrink the tubing all around the debug probe.
Make sure the tubing is equally shrunk and protects the whole PCB.

![](../../assets/debug/stlinkv3_minie_p14.jpeg)

Optionally, you may add a logo of your choice printed on paper and cut to size.
Be aware that the heat can make the ink flow a little, so you may need to experiment with what settings work with your printer.

![](../../assets/debug/stlinkv3_minie_p15.jpeg)

> [stlinkv3]: https://www.st.com/en/development-tools/stlink-v3minie.html

Reviewer: FYI this kind of linking is not used very often in the user guide. Generally we link once externally and then don't subsequently - this is reserved mostly for use in images in tables, where space is a premium. Not a big deal - the tools render it OK.

[ds009]: https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-009%20Pixhawk%20Connector%20Standard.pdf
[emdbg]: https://pypi.org/project/emdbg/
[openocd]: https://openocd.org
[stlink]: https://www.st.com/en/development-tools/stsw-link004.html

> [orbuculum]: https://github.com/orbcode/orbuculum

Reviewer: ```suggestion
```
